python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -127,13 +128,19 @@ def __init__(self, provider: NodeProvider):
         self.worker_ids: List[NodeID] = []
         # The head node (node kind "head")
         self.head_id: Optional[NodeID] = None
+        # Map of replica IDs to worker nodes in each replica
+        self.replicas_to_nodes = defaultdict(list)
 
         for node in self.all_node_ids:
-            node_kind = provider.node_tags(node)[TAG_RAY_NODE_KIND]
+            tags = provider.node_tags(node)
+            node_kind = tags[TAG_RAY_NODE_KIND]
             if node_kind == NODE_KIND_WORKER:
                 self.worker_ids.append(node)
             elif node_kind == NODE_KIND_HEAD:
                 self.head_id = node
+            if TAG_RAY_REPLICA_INDEX in tags:
+                node_replica_index = tags[TAG_RAY_REPLICA_INDEX]
+                self.replicas_to_nodes[node_replica_index].append(node)
```

Comment:
`provider.non_terminated_nodes({})` seems to have built `replicas_to_nodes`. Why do we need to build it again?

Reply:
What do you mean it's built it? I thought `provider.non_terminated_nodes({})` was all the non-terminated nodes, I don't see a mapping there between replica IDs and `non_terminated_nodes` nodes in that replica.

Reply:
See batching_node_provider.py `non_terminated_nodes` for more details.

Reply:
Done in 627fcb2.